pkgs/development/compilers/tinygo/default.nix

```diff
@@ -1,64 +1,74 @@
-{ lib, buildGoModule, fetchFromGitHub, llvm, clang-unwrapped, lld, avrgcc
-, avrdude, openocd, gcc-arm-embedded, makeWrapper, fetchurl }:
+{ lib
+, buildGoModule
+, fetchFromGitHub
+, makeWrapper
+, llvm
+, clang-unwrapped
+, lld
+, avrgcc
+, go
+, gcc-arm-embedded
+, avrdude
+, openocd
+}:
 
-let main = ./main.go;
-    gomod = ./go.mod;
-in
-buildGoModule rec {
+let
+  llvmMajor = lib.versions.major llvm.version;
+in buildGoModule rec {
   pname = "tinygo";
-  version = "0.16.0";
+  version = "0.21.0";
 
   src = fetchFromGitHub {
     owner = "tinygo-org";
-    repo = "tinygo";
+    repo = pname;
     rev = "v${version}";
-    sha256 = "063aszbsnr0myq56kms1slmrfs7m4nmg0zgh2p66lxdsifrfly7j";
+    sha256 = "sha256-ZAUvOKApwOIKWINta9jEs/M32pHgW86+a0Y1jHEMupk=";
     fetchSubmodules = true;
   };
 
-  overrideModAttrs = (_: {
-      patches = [];
-      preBuild = ''
-      rm -rf *
-      cp ${main} main.go
-      cp ${gomod} go.mod
-      chmod +w go.mod
-      '';
-  });
+  vendorSha256 = "sha256-fD9/hPRXIT7cPljA+QNWDRESqHXeCbtWHMq12280R1g=";
 
-  preBuild = "cp ${gomod} go.mod";
-
-  postBuild = "make gen-device";
+  nativeBuildInputs = [ makeWrapper ];
+  buildInputs = [ llvm clang-unwrapped ];
 
-  vendorSha256 = "12k2gin0v7aqz5543m12yhifc0xsz26qyqra5l4c68xizvzcvkxb";
+  subPackages = [ "." ];
 
   doCheck = false;
 
-  prePatch = ''
+  # We need to set GOROOT to import and compile
+  # standard go libraries with tinygo
+  allowGoReference = true;
```

Comment:
Technically no, the only thing that's needed is to have the `go` binary in the `$PATH` variable. TinyGo _usually_ figures out the GOROOT from there. But with symlinks and such, it could be mistaken.

If you don't want to set GOROOT, we could implement some way to add support for NixOS here: https://github.com/tinygo-org/tinygo/blob/release/goenv/goenv.go#L279

Reply:
We would need to hardcode the go path there which probably counts as a go reference.